src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,134 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import os
+from builtins import list, open, str
+from collections import defaultdict, namedtuple
+
+from future.moves.itertools import zip_longest
+from six import string_types
+
+from pants.reporting.reporter import Reporter
+from pants.util.dirutil import safe_mkdir
+
+
+class JsonReporter(Reporter):
+  """A reporter to capture workunit data into a JSON structure."""
+
+  _log_level_str = [
+    'FATAL',
+    'ERROR',
+    'WARN',
+    'INFO',
+    'DEBUG',
+  ]
+
+  # JSON reporting settings.
+  #   json_dir: Where the report file goes.
+  Settings = namedtuple('Settings', Reporter.Settings._fields + ('json_dir',))
+
+  def __init__(self, run_tracker, settings):
+    super(JsonReporter, self).__init__(run_tracker, settings)
+
+    # The main report output.
+    self.report_path = os.path.join(settings.json_dir, 'build.json')
+
+    # Results of the state of the build.
+    self._results = {}
+
+    # We use a stack to track the nested workunit traversal of each root.
+    self._root_id_to_workunit_stack = defaultdict(list)
+
+  def open(self):
+    """Implementation of Reporter callback."""
+
+    safe_mkdir(os.path.dirname(self.report_path))
+
+  def close(self):
+    """Implementation of Reporter callback."""
+
+    with open(self.report_path, 'w+') as fp:
+      fp.write(str(json.dumps(
```

Comment:
OK, so there's a lot going on here, hidden behind this `str` hack, which is a bit voodoo and not fully correct (or rather, only correct because of certain facts about how `json.dumps` works).

Here's the deal:

Python2 is very loose about converting character strings to byte arrays and back as needed, and it uses the interpreter's default encoding to do so. 

So in Python2 you can write bytes to a file opened in text mode or a string to a file opened in binary mode, and it works. 

But Python3 is stricter: you must write a string to a file opened in text mode, and bytes to a file opened in binary mode.

(best way to get a feel for all this is to try it in a REPL).

Now, `json.dumps()` always returns a byte array (except in some weird circumstances on Python2, which we can ignore). This is because JSON itself supports encoding all non-ASCII characters using `\uXXXX` sequences. So any valid JSON can always be represented in pure ASCII, so there is no need to worry about encodings.

When you cast it to a `str`, on Python2 this is a no-op (`dumps` already returned a byte array), and on Python3 this turns the result of `dumps` from a byte array to a character string using whatever the interpreter default encoding is, and then `fp.write()` immediately decodes it using the same default encoding. This is obviously wonky. 

Now, in practice all these encodings and decodings are no-ops, because `dumps` returned ASCII, but it's still confusing and unnecessary. It's not obvious to the casual reader what is going on, or why it works out to be correct in practice.

Instead of casting to `str`, you can just open the file in binary mode: `wb`. Then you can write the result of `dumps` directly.

Reply:
Got it - thanks for the explanation. I still ran into issues in binary mode (on Python3) using `wb` - looking back through `RunTracker` there appears to be two ways of resolving that issue - 

https://github.com/pantsbuild/pants/blob/6744c9d12ad6d9f73bac77e750a9319557d0ae94/src/python/pants/goal/run_tracker.py#L398-L403

https://github.com/pantsbuild/pants/blob/6744c9d12ad6d9f73bac77e750a9319557d0ae94/src/python/pants/goal/run_tracker.py#L432-L433

Both work, though latter appears to be what was added most recently so feels like that might be preferred - adding `safe_file_dump` to this as well looks to clear up the test failures related to `clean-all` as well.